Firestore/core/src/firebase/firestore/util/filesystem_win.cc

```diff
@@ -0,0 +1,166 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "Firestore/core/src/firebase/firestore/util/filesystem.h"
+
+#include <windows.h>
+
+#include <cerrno>
+#include <string>
+
+#include "Firestore/core/src/firebase/firestore/util/hard_assert.h"
+#include "Firestore/core/src/firebase/firestore/util/path.h"
+#include "Firestore/core/src/firebase/firestore/util/string_format.h"
+
+namespace firebase {
+namespace firestore {
+namespace util {
+
+bool Exists(const Path& path, bool* is_directory) {
+  DWORD attrs = ::GetFileAttributesW(path.c_str());
+  if (attrs == INVALID_FILE_ATTRIBUTES) {
+    return false;
+  }
+
+  if (is_directory) {
+    *is_directory = attrs & FILE_ATTRIBUTE_DIRECTORY;
+  }
+  return true;
+}
+
+Path TempDir() {
+  // Returns a null-terminated string with a trailing backslash.
+  wchar_t buffer[MAX_PATH + 1];
+  DWORD count = ::GetTempPathW(MAX_PATH, buffer);
+  HARD_ASSERT(count > 0, "Failed to determine temporary directory %s",
+              ::GetLastError());
+
+  return Path::FromUtf16(buffer, count);
+}
+
+namespace detail {
+
+Status CreateDir(const Path& path) {
+  if (::CreateDirectoryW(path.c_str(), nullptr)) {
+    return Status::OK();
+  }
+
+  DWORD error = ::GetLastError();
+  if (error == ERROR_ALREADY_EXISTS) {
+    DWORD attrs = ::GetFileAttributesW(path.c_str());
+    if (attrs == INVALID_FILE_ATTRIBUTES) {
+      error = ::GetLastError();
+
+    } else if (attrs & FILE_ATTRIBUTE_DIRECTORY) {
+      return Status::OK();
+
+    } else {
+      return Status{
+          FirestoreErrorCode::FailedPrecondition,
+          StringFormat(
+              "Could not create directory %s: non-directory already exists",
+              path.ToUtf8String())};
+    }
+  }
+
+  return Status::FromLastError(
+      error,
+      StringFormat("Could not create directory %s", path.ToUtf8String()));
+}
+
+Status DeleteDir(const Path& path) {
+  if (RemoveDirectoryW(path.c_str())) {
+    return Status::OK();
+  }
+
+  DWORD error = GetLastError();
+  if (error == ERROR_FILE_NOT_FOUND || error == ERROR_PATH_NOT_FOUND) {
+    return Status::OK();
+  }
+
+  return Status::FromLastError(
+      error,
+      StringFormat("Could not delete directory %s", path.ToUtf8String()));
+}
+
+Status DeleteFile(const Path& path) {
+  if (DeleteFileW(path.c_str())) {
+    return Status::OK();
+  }
+
+  DWORD error = GetLastError();
+  if (error == ERROR_FILE_NOT_FOUND || error == ERROR_PATH_NOT_FOUND) {
+    return Status::OK();
+  }
+
+  return Status::FromLastError(
+      error, StringFormat("Could not delete file %s", path.ToUtf8String()));
+}
+
+/**
+ * Recursively deletes a path known to be a directory.
```

Comment:
Similar to posix commit, this conflicts with the header docstring. Recommend fixing header docstring (if necessary) and removing this one.